src/test/java/io/dgraph/UpsertBlockTest.java

```diff
@@ -96,4 +96,161 @@ public void upsertBlockTest() {
     String exp3 = "{\"me\":[]}";
     assertEquals(res, exp3);
   }
+
+  @Test
+  public void upsertBlockTestMultipleUID() {
+    DgraphProto.Operation op =
+        DgraphProto.Operation.newBuilder()
+            .setSchema("email: string @index(exact) @upsert .\nname: string @index(exact) .")
+            .build();
+    dgraphClient.alter(op);
+
+    JsonArray jA = new JsonArray();
+    JsonObject p1 = new JsonObject();
+    p1.addProperty("uid", "_:alice");
+    p1.addProperty("name", "alice");
+    jA.add(p1);
+
+    JsonObject p2 = new JsonObject();
+    p2.addProperty("email", "one@dgraph.io");
+    p2.addProperty("uid", "_:alice");
+    jA.add(p2);
+
+    JsonObject p3 = new JsonObject();
+    p3.addProperty("uid", "_:bob");
+    p3.addProperty("name", "bob");
+    jA.add(p3);
+
+    JsonObject p4 = new JsonObject();
+    p4.addProperty("email", "one@dgraph.io");
+    p4.addProperty("uid", "_:bob");
+    jA.add(p4);
+
+    Mutation mu = Mutation.newBuilder().setSetJson(ByteString.copyFromUtf8(jA.toString())).build();
+    Request req = Request.newBuilder().addMutations(mu).build();
+
+    Transaction txn = dgraphClient.newTransaction();
+    txn.doRequest(req);
+    txn.commit();
+
+    String query =
+        ""
+            + "{\n"
+            + "    me(func: eq(email, \"one@dgraph.io\")) {\n"
+            + "        v as uid\n"
+            + "    }\n"
+            + "}\n";
+
+    jA = new JsonArray();
+    p1 = new JsonObject();
+    p1.addProperty("uid", "uid(v)");
+    p1.addProperty("email", "two@dgraph.io");
+    jA.add(p1);
+
+    mu = Mutation.newBuilder().setSetJson(ByteString.copyFromUtf8(jA.toString())).build();
+    req = Request.newBuilder().addMutations(mu).setQuery(query).build();
+
+    txn = dgraphClient.newTransaction();
+    txn.doRequest(req);
+    txn.commit();
+
+    String query2 =
+        ""
+            + "{\n"
+            + "    me(func: eq(email, \"two@dgraph.io\")) {\n"
+            + "        name\n"
+            + "        email\n"
+            + "    }\n"
+            + "}\n";
+
+    txn = dgraphClient.newTransaction();
+    Response response = txn.query(query2);
+    String res = response.getJson().toStringUtf8();
+    String exp =
+        "{\"me\":[{\"name\":\"alice\",\"email\":\"two@dgraph.io\"},{\"name\":\"bob\",\"email\":\"two@dgraph.io\"}]}";
```

Comment:
The order of alice and bob can change in the json response. Not the best approach for comparison